
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://apcamargo.github.io/taxopy/guide/">
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../reference/">
      
      
        
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Guide - taxopy</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CJetBrains+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"JetBrains Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../stylesheets/extra.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#guide" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="taxopy" class="md-header__button md-logo" aria-label="taxopy" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 1a2.5 2.5 0 0 0-2.5 2.5A2.5 2.5 0 0 0 11 5.79V7H7a2 2 0 0 0-2 2v.71A2.5 2.5 0 0 0 3.5 12 2.5 2.5 0 0 0 5 14.29V15H4a2 2 0 0 0-2 2v1.21A2.5 2.5 0 0 0 .5 20.5 2.5 2.5 0 0 0 3 23a2.5 2.5 0 0 0 2.5-2.5A2.5 2.5 0 0 0 4 18.21V17h4v1.21a2.5 2.5 0 0 0-1.5 2.29A2.5 2.5 0 0 0 9 23a2.5 2.5 0 0 0 2.5-2.5 2.5 2.5 0 0 0-1.5-2.29V17a2 2 0 0 0-2-2H7v-.71A2.5 2.5 0 0 0 8.5 12 2.5 2.5 0 0 0 7 9.71V9h10v.71A2.5 2.5 0 0 0 15.5 12a2.5 2.5 0 0 0 1.5 2.29V15h-1a2 2 0 0 0-2 2v1.21a2.5 2.5 0 0 0-1.5 2.29A2.5 2.5 0 0 0 15 23a2.5 2.5 0 0 0 2.5-2.5 2.5 2.5 0 0 0-1.5-2.29V17h4v1.21a2.5 2.5 0 0 0-1.5 2.29A2.5 2.5 0 0 0 21 23a2.5 2.5 0 0 0 2.5-2.5 2.5 2.5 0 0 0-1.5-2.29V17a2 2 0 0 0-2-2h-1v-.71A2.5 2.5 0 0 0 20.5 12 2.5 2.5 0 0 0 19 9.71V9a2 2 0 0 0-2-2h-4V5.79a2.5 2.5 0 0 0 1.5-2.29A2.5 2.5 0 0 0 12 1m0 1.5a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1M6 11a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m12 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1M3 19.5a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m6 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m6 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m6 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            taxopy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Guide
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M7.5 2c-1.79 1.15-3 3.18-3 5.5s1.21 4.35 3.03 5.5C4.46 13 2 10.54 2 7.5A5.5 5.5 0 0 1 7.5 2m11.57 1.5 1.43 1.43L4.93 20.5 3.5 19.07zm-6.18 2.43L11.41 5 9.97 6l.42-1.7L9 3.24l1.75-.12.58-1.65L12 3.1l1.73.03-1.35 1.13zm-3.3 3.61-1.16-.73-1.12.78.34-1.32-1.09-.83 1.36-.09.45-1.29.51 1.27 1.36.03-1.05.87zM19 13.5a5.5 5.5 0 0 1-5.5 5.5c-1.22 0-2.35-.4-3.26-1.07l7.69-7.69c.67.91 1.07 2.04 1.07 3.26m-4.4 6.58 2.77-1.15-.24 3.35zm4.33-2.7 1.15-2.77 2.2 2.54zm1.15-4.96-1.14-2.78 3.34.24zM9.63 18.93l2.77 1.15-2.53 2.19z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="yellow" data-md-color-accent="amber"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="yellow"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/apcamargo/taxopy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    apcamargo/taxopy
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
  
    
  
  Guide

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../reference/" class="md-tabs__link">
        
  
  
    
  
  Reference

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="taxopy" class="md-nav__button md-logo" aria-label="taxopy" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 1a2.5 2.5 0 0 0-2.5 2.5A2.5 2.5 0 0 0 11 5.79V7H7a2 2 0 0 0-2 2v.71A2.5 2.5 0 0 0 3.5 12 2.5 2.5 0 0 0 5 14.29V15H4a2 2 0 0 0-2 2v1.21A2.5 2.5 0 0 0 .5 20.5 2.5 2.5 0 0 0 3 23a2.5 2.5 0 0 0 2.5-2.5A2.5 2.5 0 0 0 4 18.21V17h4v1.21a2.5 2.5 0 0 0-1.5 2.29A2.5 2.5 0 0 0 9 23a2.5 2.5 0 0 0 2.5-2.5 2.5 2.5 0 0 0-1.5-2.29V17a2 2 0 0 0-2-2H7v-.71A2.5 2.5 0 0 0 8.5 12 2.5 2.5 0 0 0 7 9.71V9h10v.71A2.5 2.5 0 0 0 15.5 12a2.5 2.5 0 0 0 1.5 2.29V15h-1a2 2 0 0 0-2 2v1.21a2.5 2.5 0 0 0-1.5 2.29A2.5 2.5 0 0 0 15 23a2.5 2.5 0 0 0 2.5-2.5 2.5 2.5 0 0 0-1.5-2.29V17h4v1.21a2.5 2.5 0 0 0-1.5 2.29A2.5 2.5 0 0 0 21 23a2.5 2.5 0 0 0 2.5-2.5 2.5 2.5 0 0 0-1.5-2.29V17a2 2 0 0 0-2-2h-1v-.71A2.5 2.5 0 0 0 20.5 12 2.5 2.5 0 0 0 19 9.71V9a2 2 0 0 0-2-2h-4V5.79a2.5 2.5 0 0 0 1.5-2.29A2.5 2.5 0 0 0 12 1m0 1.5a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1M6 11a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m12 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1M3 19.5a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m6 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m6 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1m6 0a1 1 0 0 1 1 1 1 1 0 0 1-1 1 1 1 0 0 1-1-1 1 1 0 0 1 1-1"/></svg>

    </a>
    taxopy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/apcamargo/taxopy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    apcamargo/taxopy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Guide
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Guide
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#initializing-a-taxonomy-database-in-a-taxdb-object" class="md-nav__link">
    <span class="md-ellipsis">
      
        Initializing a taxonomy database in a TaxDb object
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-taxon-object" class="md-nav__link">
    <span class="md-ellipsis">
      
        The Taxon object
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#identifying-taxa-shared-between-multiple-lineages" class="md-nav__link">
    <span class="md-ellipsis">
      
        Identifying taxa shared between multiple lineages
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Identifying taxa shared between multiple lineages">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#lowest-common-ancestor" class="md-nav__link">
    <span class="md-ellipsis">
      
        Lowest common ancestor
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#majority-vote" class="md-nav__link">
    <span class="md-ellipsis">
      
        Majority vote
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Majority vote">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-fraction-parameter" class="md-nav__link">
    <span class="md-ellipsis">
      
        The fraction parameter
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#the-weights-parameter" class="md-nav__link">
    <span class="md-ellipsis">
      
        The weights parameter
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#retrieving-taxids-from-taxa-names" class="md-nav__link">
    <span class="md-ellipsis">
      
        Retrieving TaxIds from taxa names
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Retrieving TaxIds from taxa names">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#retrieval-of-taxa-with-nearly-matching-names-though-fuzzy-search" class="md-nav__link">
    <span class="md-ellipsis">
      
        Retrieval of taxa with nearly matching names though fuzzy search
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Reference
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="guide">Guide<a class="headerlink" href="#guide" title="Permanent link">#</a></h1>
<h2 id="initializing-a-taxonomy-database-in-a-taxdb-object">Initializing a taxonomy database in a <code>TaxDb</code> object<a class="headerlink" href="#initializing-a-taxonomy-database-in-a-taxdb-object" title="Permanent link">#</a></h2>
<p>To get started, you need to create a <a class="autorefs autorefs-internal" title="            taxopy.TaxDb" href="../reference/#taxopy.TaxDb"><code>TaxDb</code></a> object, which will store data related to the taxonomic database, such as taxonomic identifiers (or TaxIds), names, and hierarchies. This can be achieved by downloading the set of files that store this data, known as <a href="https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump_readme.txt">taxdump</a>, from an online source (NCBI by default) or by providing your own taxdump files.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:3"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><input id="__tabbed_1_3" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Downloading from NCBI</label><label for="__tabbed_1_2">Downloading from a different source</label><label for="__tabbed_1_3">Local taxdump</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">taxopy</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">taxdb</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">TaxDb</span><span class="p">()</span><span class="c1"># (1)!</span>
</span></code></pre></div>
<ol>
<li>By default, <code>taxopy</code> deletes the taxdump files after creating the object. To retain the files in the working directory, set <code>keep_files=True</code>. You can specify the directory where the files are stored using the <code>taxdb_dir</code> parameter.</li>
</ol>
</div>
<div class="tabbed-block">
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">taxopy</span>
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://github.com/shenwei356/gtdb-taxdump/releases/download/v0.5.0/gtdb-taxdump-R220.tar.gz&quot;</span>
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">taxdb</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">TaxDb</span><span class="p">(</span>
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="gp">... </span>    <span class="n">taxdump_url</span><span class="o">=</span><span class="n">url</span><span class="c1"># (1)!</span>
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="gp">... </span><span class="p">)</span>
</span></code></pre></div>
<ol>
<li>The <code>taxdump_url</code> parameter is used to specify the URL of the taxdump file to download. In this case, we are using a <a href="https://github.com/shenwei356/gtdb-taxdump">custom a GTDB taxdump provided by Wei Shen</a>.</li>
</ol>
</div>
<div class="tabbed-block">
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">taxopy</span>
</span><span id="__span-2-2"><a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">taxdb</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">TaxDb</span><span class="p">(</span>
</span><span id="__span-2-3"><a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a><span class="gp">... </span>    <span class="n">nodes_dmp</span><span class="o">=</span><span class="s2">&quot;taxdb/nodes.dmp&quot;</span><span class="p">,</span>
</span><span id="__span-2-4"><a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a><span class="gp">... </span>    <span class="n">names_dmp</span><span class="o">=</span><span class="s2">&quot;taxdb/names.dmp&quot;</span><span class="p">,</span>
</span><span id="__span-2-5"><a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a><span class="gp">... </span>    <span class="n">merged_dmp</span><span class="o">=</span><span class="s2">&quot;taxdb/merged.dmp&quot;</span><span class="p">,</span><span class="c1"># (1)!</span>
</span><span id="__span-2-6"><a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a><span class="gp">... </span><span class="p">)</span>
</span></code></pre></div>
<ol>
<li>The <code>merged_dmp</code> parameter is optional. However, if you want to support legacy TaxIds (those merged into other identifiers), you need to provide a <code>merged.dmp</code> file. This is not necessary if the data is downloaded from an online source, which will include its own merged.dmp file.</li>
</ol>
</div>
</div>
</div>
<p>The <a class="autorefs autorefs-internal" title="            taxopy.TaxDb" href="../reference/#taxopy.TaxDb"><code>TaxDb</code></a> object stores the names, ranks, and parent-child relationships of all taxa, each represented by their respective TaxIds. For instance, <a href="https://www.ncbi.nlm.nih.gov/datasets/taxonomy/2/">TaxId 2</a> corresponds to the "<em>Bacteria</em>" taxon, which has the rank of superkingdom. The parent taxon of <em>Bacteria</em> is <a href="https://www.ncbi.nlm.nih.gov/datasets/taxonomy/131567/">TaxId 131567</a>, corresponding to "<em>cellular organisms</em>".</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxdb</span><span class="o">.</span><span class="n">taxid2name</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span id="__span-3-2"><a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a><span class="go">Bacteria</span>
</span><span id="__span-3-3"><a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxdb</span><span class="o">.</span><span class="n">taxid2rank</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span id="__span-3-4"><a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a><span class="go">superkingdom</span>
</span><span id="__span-3-5"><a id="__codelineno-3-5" name="__codelineno-3-5" href="#__codelineno-3-5"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxdb</span><span class="o">.</span><span class="n">taxid2parent</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
</span><span id="__span-3-6"><a id="__codelineno-3-6" name="__codelineno-3-6" href="#__codelineno-3-6"></a><span class="go">131567</span>
</span></code></pre></div>
<p>To retrieve the TaxId to which a legacy TaxId has been merged, you can use the <code>oldtaxid2newtaxid</code> attribute. For instance, the legacy TaxIds 260 and 29537 have been merged into <a href="https://www.ncbi.nlm.nih.gov/datasets/taxonomy/143224/">TaxId 143224</a> ("<em>Zobellia uliginosa</em>").</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxdb</span><span class="o">.</span><span class="n">oldtaxid2newtaxid</span><span class="p">[</span><span class="mi">260</span><span class="p">])</span>
</span><span id="__span-4-2"><a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a><span class="go">143224</span>
</span><span id="__span-4-3"><a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxdb</span><span class="o">.</span><span class="n">oldtaxid2newtaxid</span><span class="p">[</span><span class="mi">29537</span><span class="p">])</span>
</span><span id="__span-4-4"><a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a><span class="go">143224</span>
</span></code></pre></div>
<h2 id="the-taxon-object">The <code>Taxon</code> object<a class="headerlink" href="#the-taxon-object" title="Permanent link">#</a></h2>
<p><a class="autorefs autorefs-internal" title="            taxopy.Taxon" href="../reference/#taxopy.Taxon"><code>Taxon</code></a> objects represent individual taxa within the taxonomy database. These objects are initialized using a TaxId and a corresponding <a class="autorefs autorefs-internal" title="            taxopy.TaxDb" href="../reference/#taxopy.TaxDb"><code>TaxDb</code></a> object, from which the taxon data is retrieved.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">saccharomyces</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="mi">4930</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span>
</span><span id="__span-5-2"><a id="__codelineno-5-2" name="__codelineno-5-2" href="#__codelineno-5-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">human</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="mi">9606</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span>
</span><span id="__span-5-3"><a id="__codelineno-5-3" name="__codelineno-5-3" href="#__codelineno-5-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">gorilla</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="mi">9593</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span>
</span><span id="__span-5-4"><a id="__codelineno-5-4" name="__codelineno-5-4" href="#__codelineno-5-4"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">lagomorpha</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="mi">9975</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span>
</span></code></pre></div>
<p>Each <a class="autorefs autorefs-internal" title="            taxopy.Taxon" href="../reference/#taxopy.Taxon"><code>Taxon</code></a> object stores various data related to the taxon, including its TaxId, names, rank, and lineage. The lineage data comprises the TaxIds, scientific names, and ranks of its parent taxa.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">taxid</span><span class="p">)</span>
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="go">9975</span>
</span><span id="__span-6-3"><a id="__codelineno-6-3" name="__codelineno-6-3" href="#__codelineno-6-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">taxid_lineage</span><span class="p">)</span>
</span><span id="__span-6-4"><a id="__codelineno-6-4" name="__codelineno-6-4" href="#__codelineno-6-4"></a><span class="go">[9975, 314147, 314146, 1437010, 9347, 32525, 40674, 32524, 32523, 1338369, 8287, 117571, 117570, 7776, 7742, 89593, 7711, 33511, 33213, 6072, 33208, 33154, 2759, 131567, 1]</span>
</span><span id="__span-6-5"><a id="__codelineno-6-5" name="__codelineno-6-5" href="#__codelineno-6-5"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">ranked_taxid_lineage</span><span class="p">)</span>
</span><span id="__span-6-6"><a id="__codelineno-6-6" name="__codelineno-6-6" href="#__codelineno-6-6"></a><span class="go">[(&#39;order&#39;, 9975), (&#39;clade&#39;, 314147), (&#39;superorder&#39;, 314146), (&#39;clade&#39;, 1437010), (&#39;clade&#39;, 9347), (&#39;clade&#39;, 32525), (&#39;class&#39;, 40674), (&#39;clade&#39;, 32524), (&#39;clade&#39;, 32523), (&#39;clade&#39;, 1338369), (&#39;superclass&#39;, 8287), (&#39;clade&#39;, 117571), (&#39;clade&#39;, 117570), (&#39;clade&#39;, 7776), (&#39;clade&#39;, 7742), (&#39;subphylum&#39;, 89593), (&#39;phylum&#39;, 7711), (&#39;clade&#39;, 33511), (&#39;clade&#39;, 33213), (&#39;clade&#39;, 6072), (&#39;kingdom&#39;, 33208), (&#39;clade&#39;, 33154), (&#39;superkingdom&#39;, 2759), (&#39;no rank&#39;, 131567), (&#39;no rank&#39;, 1)]</span>
</span><span id="__span-6-7"><a id="__codelineno-6-7" name="__codelineno-6-7" href="#__codelineno-6-7"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-6-8"><a id="__codelineno-6-8" name="__codelineno-6-8" href="#__codelineno-6-8"></a><span class="go">Lagomorpha</span>
</span><span id="__span-6-9"><a id="__codelineno-6-9" name="__codelineno-6-9" href="#__codelineno-6-9"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">all_names</span><span class="p">)</span>
</span><span id="__span-6-10"><a id="__codelineno-6-10" name="__codelineno-6-10" href="#__codelineno-6-10"></a><span class="go">{&#39;authority&#39;: [&#39;Lagomorpha Brandt, 1855&#39;], &#39;scientific name&#39;: [&#39;Lagomorpha&#39;], &#39;blast name&#39;: [&#39;rabbits &amp; hares&#39;]}</span>
</span><span id="__span-6-11"><a id="__codelineno-6-11" name="__codelineno-6-11" href="#__codelineno-6-11"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">name_lineage</span><span class="p">)</span>
</span><span id="__span-6-12"><a id="__codelineno-6-12" name="__codelineno-6-12" href="#__codelineno-6-12"></a><span class="go">[&#39;Lagomorpha&#39;, &#39;Glires&#39;, &#39;Euarchontoglires&#39;, &#39;Boreoeutheria&#39;, &#39;Eutheria&#39;, &#39;Theria&#39;, &#39;Mammalia&#39;, &#39;Amniota&#39;, &#39;Tetrapoda&#39;, &#39;Dipnotetrapodomorpha&#39;, &#39;Sarcopterygii&#39;, &#39;Euteleostomi&#39;, &#39;Teleostomi&#39;, &#39;Gnathostomata&#39;, &#39;Vertebrata&#39;, &#39;Craniata&#39;, &#39;Chordata&#39;, &#39;Deuterostomia&#39;, &#39;Bilateria&#39;, &#39;Eumetazoa&#39;, &#39;Metazoa&#39;, &#39;Opisthokonta&#39;, &#39;Eukaryota&#39;, &#39;cellular organisms&#39;, &#39;root&#39;]</span>
</span><span id="__span-6-13"><a id="__codelineno-6-13" name="__codelineno-6-13" href="#__codelineno-6-13"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">ranked_name_lineage</span><span class="p">)</span>
</span><span id="__span-6-14"><a id="__codelineno-6-14" name="__codelineno-6-14" href="#__codelineno-6-14"></a><span class="go">[(&#39;order&#39;, &#39;Lagomorpha&#39;), (&#39;clade&#39;, &#39;Glires&#39;), (&#39;superorder&#39;, &#39;Euarchontoglires&#39;), (&#39;clade&#39;, &#39;Boreoeutheria&#39;), (&#39;clade&#39;, &#39;Eutheria&#39;), (&#39;clade&#39;, &#39;Theria&#39;), (&#39;class&#39;, &#39;Mammalia&#39;), (&#39;clade&#39;, &#39;Amniota&#39;), (&#39;clade&#39;, &#39;Tetrapoda&#39;), (&#39;clade&#39;, &#39;Dipnotetrapodomorpha&#39;), (&#39;superclass&#39;, &#39;Sarcopterygii&#39;), (&#39;clade&#39;, &#39;Euteleostomi&#39;), (&#39;clade&#39;, &#39;Teleostomi&#39;), (&#39;clade&#39;, &#39;Gnathostomata&#39;), (&#39;clade&#39;, &#39;Vertebrata&#39;), (&#39;subphylum&#39;, &#39;Craniata&#39;), (&#39;phylum&#39;, &#39;Chordata&#39;), (&#39;clade&#39;, &#39;Deuterostomia&#39;), (&#39;clade&#39;, &#39;Bilateria&#39;), (&#39;clade&#39;, &#39;Eumetazoa&#39;), (&#39;kingdom&#39;, &#39;Metazoa&#39;), (&#39;clade&#39;, &#39;Opisthokonta&#39;), (&#39;superkingdom&#39;, &#39;Eukaryota&#39;), (&#39;no rank&#39;, &#39;cellular organisms&#39;), (&#39;no rank&#39;, &#39;root&#39;)]</span>
</span><span id="__span-6-15"><a id="__codelineno-6-15" name="__codelineno-6-15" href="#__codelineno-6-15"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">rank_name_dictionary</span><span class="p">)</span>
</span><span id="__span-6-16"><a id="__codelineno-6-16" name="__codelineno-6-16" href="#__codelineno-6-16"></a><span class="go">OrderedDict({&#39;order&#39;: &#39;Lagomorpha&#39;, &#39;clade&#39;: &#39;Opisthokonta&#39;, &#39;superorder&#39;: &#39;Euarchontoglires&#39;, &#39;class&#39;: &#39;Mammalia&#39;, &#39;superclass&#39;: &#39;Sarcopterygii&#39;, &#39;subphylum&#39;: &#39;Craniata&#39;, &#39;phylum&#39;: &#39;Chordata&#39;, &#39;kingdom&#39;: &#39;Metazoa&#39;, &#39;superkingdom&#39;: &#39;Eukaryota&#39;})</span>
</span><span id="__span-6-17"><a id="__codelineno-6-17" name="__codelineno-6-17" href="#__codelineno-6-17"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">lagomorpha</span><span class="o">.</span><span class="n">rank</span><span class="p">)</span>
</span><span id="__span-6-18"><a id="__codelineno-6-18" name="__codelineno-6-18" href="#__codelineno-6-18"></a><span class="go">order</span>
</span></code></pre></div>
<p>A taxon's scientific name is stored in the <code>name</code> field, while <code>all_names</code> is a dictionary from the kind of name to a list of names. </p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">sus</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="mi">9823</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span>
</span><span id="__span-7-2"><a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sus</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-7-3"><a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a><span class="go">&#39;Sus scrofa&#39;</span>
</span><span id="__span-7-4"><a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sus</span><span class="o">.</span><span class="n">all_names</span><span class="p">)</span>
</span><span id="__span-7-5"><a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a><span class="go">{&#39;genbank common name&#39;: [&#39;pig&#39;], &#39;common name&#39;: [&#39;pigs&#39;, &#39;swine&#39;, &#39;wild boar&#39;], &#39;authority&#39;: [&#39;Sus scrofa Linnaeus, 1758&#39;], &#39;includes&#39;: [&#39;Sus scrofa LW&#39;, &#39;Sus scrofa Pietrain&#39;], &#39;scientific name&#39;: [&#39;Sus scrofa&#39;]}</span>
</span><span id="__span-7-6"><a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">sus</span><span class="o">.</span><span class="n">all_names</span><span class="p">[</span><span class="s1">&#39;common name&#39;</span><span class="p">])</span>
</span><span id="__span-7-7"><a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a><span class="go">[&#39;pigs&#39;, &#39;swine&#39;, &#39;wild boar&#39;]</span>
</span></code></pre></div>
<p>To obtain the <a class="autorefs autorefs-internal" title="            taxopy.Taxon" href="../reference/#taxopy.Taxon"><code>Taxon</code></a> object for the parent of a specified taxon, you can use the <code>parent</code> method.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">saccharomyces_parent</span> <span class="o">=</span> <span class="n">saccharomyces</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">taxdb</span><span class="p">)</span>
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">saccharomyces_parent</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-8-3"><a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a><span class="go">Saccharomycetaceae</span>
</span><span id="__span-8-4"><a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">saccharomyces_parent</span><span class="o">.</span><span class="n">rank</span><span class="p">)</span>
</span><span id="__span-8-5"><a id="__codelineno-8-5" name="__codelineno-8-5" href="#__codelineno-8-5"></a><span class="go">family</span>
</span></code></pre></div>
<h2 id="identifying-taxa-shared-between-multiple-lineages">Identifying taxa shared between multiple lineages<a class="headerlink" href="#identifying-taxa-shared-between-multiple-lineages" title="Permanent link">#</a></h2>
<p>Due to the hierarchical nature of taxonomies, different taxa may share common ancestors at one or more ranks in their lineages. For example, the <em>Lagomorpha</em> and <em>Homo sapiens</em> lineages have common taxa from the root of the taxonomy (parent to all other taxa) up until <em>Euarchontoglires</em>, from which they then diverge.</p>
<p><pre class="mermaid"><code>---
config:
  layout: elk
---
graph LR
    subgraph cluster ["*Lagomorpha*, *Homo sapiens*"]
        R(Root) -.-&gt; E(*Euarchontoglires*)
    end
    E --&gt; P(*Primates*)
    P -.-&gt; H(*Homo sapiens*)
    E --&gt; G(*Glires*)
    G --&gt; L(*Lagomorpha*)</code></pre>
In <code>taxopy</code>, <a class="autorefs autorefs-internal" title="            taxopy.Taxon" href="../reference/#taxopy.Taxon"><code>Taxon</code></a> objects store the lineage data for each taxon, enabling the identification of shared taxa between two or more lineages. <code>taxopy</code> provides two functions for this purpose:</p>
<ul>
<li><a class="autorefs autorefs-internal" title="            taxopy.find_lca" href="../reference/#taxopy.find_lca"><code>find_lca</code></a>: Identifies the most specific taxon shared among all provided lineages.</li>
<li><a class="autorefs autorefs-internal" title="            taxopy.find_majority_vote" href="../reference/#taxopy.find_majority_vote"><code>find_majority_vote</code></a>: Identifies the most specific taxon common to more than a specified fraction of the provided lineages (default is more than 50%).</li>
</ul>
<h3 id="lowest-common-ancestor">Lowest common ancestor<a class="headerlink" href="#lowest-common-ancestor" title="Permanent link">#</a></h3>
<p>The lowest common ancestor (LCA) of a set of taxa is the most specific taxon that is common to all the lineages. In the example above, the LCA between <em>Lagomorpha</em> and <em>Homo sapiens</em> is the <em>Euarchontoglires</em> superorder. This can be easily determined using <code>taxopy</code> through the <a class="autorefs autorefs-internal" title="            taxopy.find_lca" href="../reference/#taxopy.find_lca"><code>find_lca</code></a>.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">find_lca</span><span class="p">([</span><span class="n">lagomorpha</span><span class="p">,</span> <span class="n">human</span><span class="p">],</span> <span class="n">taxdb</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-9-2"><a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a><span class="go">Euarchontoglires</span>
</span></code></pre></div>
<h3 id="majority-vote">Majority vote<a class="headerlink" href="#majority-vote" title="Permanent link">#</a></h3>
<p>The LCA can be uninformative when determining a representative taxon for a set of lineages. For instance, consider the lineages of <em>Lagomorpha</em>, <em>Homo sapiens</em>, and <em>Saccharomyces</em>.</p>
<pre class="mermaid"><code>---
config:
  layout: elk
---
graph LR
    subgraph cluster1 ["*Saccharomyces*, *Lagomorpha*, *Homo sapiens*"]
        R(Root) -.-&gt; O(*Opisthokonta*) --&gt; M(*Metazoa*)
        subgraph cluster2 ["*Lagomorpha*, *Homo sapiens*"]
            M -.-&gt; E(*Euarchontoglires*)
        end
    end
    E --&gt; P(*Primates*)
    P -.-&gt; H(*Homo sapiens*)
    E --&gt; G(*Glires*)
    G --&gt; L(*Lagomorpha*)
    O --&gt; F(*Fungi*) -...-&gt; S(*Saccharomyces*)</code></pre>
<p><em>Saccharomyces</em> is very distantly related to both <em>Lagomorpha</em> and <em>Homo sapiens</em>, making the LCA of these three lineages <em>Opisthokonta</em>, a broad group within <em>Eukaryota</em> that includes both animals and fungi.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">find_lca</span><span class="p">([</span><span class="n">saccharomyces</span><span class="p">,</span> <span class="n">lagomorpha</span><span class="p">,</span> <span class="n">human</span><span class="p">],</span> <span class="n">taxdb</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-10-2"><a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="go">Opisthokonta</span>
</span></code></pre></div>
<p>If you want to find a taxon that represents most of the lineages in a set, the LCA might be too broad if a single distantly related organism is included, which is especially common when working with a large set of lineages. Instead, you can use the <a class="autorefs autorefs-internal" title="            taxopy.find_majority_vote" href="../reference/#taxopy.find_majority_vote"><code>find_majority_vote</code></a> function to determine the taxon shared by <em>most</em> of the lineages, but not necessarily <em>all</em> of them. This approach generally yields a more specific taxon than the LCA. In the example above, <em>Euarchontoglires</em> is the most specific taxon shared by more than half of the three lineages.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span>
</span><span id="__span-11-2"><a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a><span class="gp">... </span>    <span class="n">taxopy</span><span class="o">.</span><span class="n">find_majority_vote</span><span class="p">(</span>
</span><span id="__span-11-3"><a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="gp">... </span>        <span class="p">[</span><span class="n">saccharomyces</span><span class="p">,</span> <span class="n">lagomorpha</span><span class="p">,</span> <span class="n">human</span><span class="p">],</span> <span class="n">taxdb</span>
</span><span id="__span-11-4"><a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="gp">... </span>    <span class="p">)</span><span class="o">.</span><span class="n">name</span>
</span><span id="__span-11-5"><a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a><span class="gp">... </span><span class="p">)</span>
</span><span id="__span-11-6"><a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a><span class="go">Euarchontoglires</span>
</span></code></pre></div>
<h4 id="the-fraction-parameter">The <code>fraction</code> parameter<a class="headerlink" href="#the-fraction-parameter" title="Permanent link">#</a></h4>
<p>By default, <a class="autorefs autorefs-internal" title="            taxopy.find_majority_vote" href="../reference/#taxopy.find_majority_vote"><code>find_majority_vote</code></a> requires the taxon to be shared by more than half of the lineages. You can adjust this threshold by setting the <code>fraction</code> parameter to a value greater than <code>0</code> and less than <code>1</code> (default is <code>0.5</code>). Higher values will require the resulting taxon to be shared by more lineages, approaching the LCA.</p>
<p>Let's add <em>Gorilla gorilla</em> to the set of lineages we are investigating.</p>
<pre class="mermaid"><code>---
config:
  layout: elk
---
graph LR
    subgraph cluster1 ["*Saccharomyces*, *Lagomorpha*, *Homo sapiens*, *Gorilla gorilla*"]
        R(Root) -.-&gt; O(*Opisthokonta*) --&gt; M(*Metazoa*)
        subgraph cluster2 ["*Lagomorpha*, *Homo sapiens*, *Gorilla gorilla*"]
            M -.-&gt; E(*Euarchontoglires*)
            E -.-&gt; HN(*Homininae*)
            subgraph cluster3 ["*Homo sapiens*, *Gorilla gorilla*"]
                HN
            end
        end
    end
    HN -.-&gt; H(*Homo sapiens*)
    HN -.-&gt; GG(*Gorilla gorilla*)
    E --&gt; G(*Glires*)
    G --&gt; L(*Lagomorpha*)
    O --&gt; F(*Fungi*) -...-&gt; S(*Saccharomyces*)</code></pre>
<p>If we consider <em>Lagomorpha</em>, <em>Homo sapiens</em>, and <em>Gorilla gorilla</em> (ignoring <em>Saccharomyces</em> for now), the most specific taxon that is shared by more than half of these three taxa is <em>Homininae</em>, which is common to two of them. However, if we want to determine the most specific taxon that is common to more than two-thirds of the lineages, we can change the value of <code>fraction</code> to <code>0.67</code>. This will return <em>Euarchontoglires</em>, which, in this particular case, is also the LCA of the three lineages.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">find_majority_vote</span><span class="p">([</span><span class="n">lagomorpha</span><span class="p">,</span> <span class="n">human</span><span class="p">,</span> <span class="n">gorilla</span><span class="p">],</span> <span class="n">taxdb</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="go">Homininae</span>
</span><span id="__span-12-3"><a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span>
</span><span id="__span-12-4"><a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a><span class="gp">... </span>    <span class="n">taxopy</span><span class="o">.</span><span class="n">find_majority_vote</span><span class="p">(</span>
</span><span id="__span-12-5"><a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a><span class="gp">... </span>        <span class="p">[</span><span class="n">lagomorpha</span><span class="p">,</span> <span class="n">human</span><span class="p">,</span> <span class="n">gorilla</span><span class="p">],</span> <span class="n">taxdb</span><span class="p">,</span> <span class="n">fraction</span><span class="o">=</span><span class="mf">0.67</span>
</span><span id="__span-12-6"><a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a><span class="gp">... </span>    <span class="p">)</span><span class="o">.</span><span class="n">name</span>
</span><span id="__span-12-7"><a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a><span class="gp">... </span><span class="p">)</span>
</span><span id="__span-12-8"><a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a><span class="go">Euarchontoglires</span>
</span></code></pre></div>
<p>The level of agreement among the lineages used to find the common taxon is stored in the <code>agreement</code> attribute. This is computed as the ratio of the number of lineages in which the shared taxon is present to the total number of lineages used.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span>
</span><span id="__span-13-2"><a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a><span class="gp">... </span>    <span class="n">taxopy</span><span class="o">.</span><span class="n">find_majority_vote</span><span class="p">(</span>
</span><span id="__span-13-3"><a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="gp">... </span>        <span class="p">[</span><span class="n">lagomorpha</span><span class="p">,</span> <span class="n">human</span><span class="p">,</span> <span class="n">gorilla</span><span class="p">],</span> <span class="n">taxdb</span>
</span><span id="__span-13-4"><a id="__codelineno-13-4" name="__codelineno-13-4" href="#__codelineno-13-4"></a><span class="gp">... </span>    <span class="p">)</span><span class="o">.</span><span class="n">agreement</span>
</span><span id="__span-13-5"><a id="__codelineno-13-5" name="__codelineno-13-5" href="#__codelineno-13-5"></a><span class="gp">... </span><span class="p">)</span>
</span><span id="__span-13-6"><a id="__codelineno-13-6" name="__codelineno-13-6" href="#__codelineno-13-6"></a><span class="go">0.6666666666666666</span>
</span><span id="__span-13-7"><a id="__codelineno-13-7" name="__codelineno-13-7" href="#__codelineno-13-7"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span>
</span><span id="__span-13-8"><a id="__codelineno-13-8" name="__codelineno-13-8" href="#__codelineno-13-8"></a><span class="gp">... </span>    <span class="n">taxopy</span><span class="o">.</span><span class="n">find_majority_vote</span><span class="p">(</span>
</span><span id="__span-13-9"><a id="__codelineno-13-9" name="__codelineno-13-9" href="#__codelineno-13-9"></a><span class="gp">... </span>        <span class="p">[</span><span class="n">lagomorpha</span><span class="p">,</span> <span class="n">human</span><span class="p">,</span> <span class="n">gorilla</span><span class="p">],</span> <span class="n">taxdb</span><span class="p">,</span> <span class="n">fraction</span><span class="o">=</span><span class="mf">0.67</span>
</span><span id="__span-13-10"><a id="__codelineno-13-10" name="__codelineno-13-10" href="#__codelineno-13-10"></a><span class="gp">... </span>    <span class="p">)</span><span class="o">.</span><span class="n">agreement</span>
</span><span id="__span-13-11"><a id="__codelineno-13-11" name="__codelineno-13-11" href="#__codelineno-13-11"></a><span class="gp">... </span><span class="p">)</span>
</span><span id="__span-13-12"><a id="__codelineno-13-12" name="__codelineno-13-12" href="#__codelineno-13-12"></a><span class="go">1.0</span>
</span></code></pre></div>
<h4 id="the-weights-parameter">The <code>weights</code> parameter<a class="headerlink" href="#the-weights-parameter" title="Permanent link">#</a></h4>
<p><a class="autorefs autorefs-internal" title="            taxopy.find_majority_vote" href="../reference/#taxopy.find_majority_vote"><code>find_majority_vote</code></a> also allows you to assign different weights to taxa when determining the shared taxon using the <code>weights</code> parameter. For example, the most specific taxon common to more than half of the four taxa in the figure above is <em>Euarchontoglires</em> (shared by three out of the four). However, if we set the weight of <em>Saccharomyces</em> to <code>3</code> and the weights of the remaining lineages to <code>1</code> each, the function will return <em>Opisthokonta</em>, effectively balancing the total weight of <em>Saccharomyces</em> with the three <em>Euarchontoglires</em> taxa.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span>
</span><span id="__span-14-2"><a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a><span class="gp">... </span>    <span class="n">taxopy</span><span class="o">.</span><span class="n">find_majority_vote</span><span class="p">(</span>
</span><span id="__span-14-3"><a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a><span class="gp">... </span>        <span class="p">[</span><span class="n">saccharomyces</span><span class="p">,</span> <span class="n">human</span><span class="p">,</span> <span class="n">gorilla</span><span class="p">,</span> <span class="n">lagomorpha</span><span class="p">],</span> <span class="n">taxdb</span>
</span><span id="__span-14-4"><a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a><span class="gp">... </span>    <span class="p">)</span><span class="o">.</span><span class="n">name</span>
</span><span id="__span-14-5"><a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a><span class="gp">... </span><span class="p">)</span>
</span><span id="__span-14-6"><a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a><span class="go">Euarchontoglires</span>
</span><span id="__span-14-7"><a id="__codelineno-14-7" name="__codelineno-14-7" href="#__codelineno-14-7"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span>
</span><span id="__span-14-8"><a id="__codelineno-14-8" name="__codelineno-14-8" href="#__codelineno-14-8"></a><span class="gp">... </span>    <span class="n">taxopy</span><span class="o">.</span><span class="n">find_majority_vote</span><span class="p">(</span>
</span><span id="__span-14-9"><a id="__codelineno-14-9" name="__codelineno-14-9" href="#__codelineno-14-9"></a><span class="gp">... </span>        <span class="p">[</span><span class="n">saccharomyces</span><span class="p">,</span> <span class="n">human</span><span class="p">,</span> <span class="n">gorilla</span><span class="p">,</span> <span class="n">lagomorpha</span><span class="p">],</span>
</span><span id="__span-14-10"><a id="__codelineno-14-10" name="__codelineno-14-10" href="#__codelineno-14-10"></a><span class="gp">... </span>        <span class="n">taxdb</span><span class="p">,</span>
</span><span id="__span-14-11"><a id="__codelineno-14-11" name="__codelineno-14-11" href="#__codelineno-14-11"></a><span class="gp">... </span>        <span class="n">weights</span><span class="o">=</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
</span><span id="__span-14-12"><a id="__codelineno-14-12" name="__codelineno-14-12" href="#__codelineno-14-12"></a><span class="gp">... </span>    <span class="p">)</span><span class="o">.</span><span class="n">name</span>
</span><span id="__span-14-13"><a id="__codelineno-14-13" name="__codelineno-14-13" href="#__codelineno-14-13"></a><span class="gp">... </span><span class="p">)</span>
</span><span id="__span-14-14"><a id="__codelineno-14-14" name="__codelineno-14-14" href="#__codelineno-14-14"></a><span class="go">Opisthokonta</span>
</span></code></pre></div>
<p>This functionality is useful in various scenarios. For example, if you have a set of taxa with different numbers of individuals in a community, you can use the number of individuals of each taxon as weights to determine the most specific taxon common to more than half of the individuals. Another popular application is the taxonomic assignment of metagenomic sequences<sup id="fnref:1"><a class="footnote-ref" href="#fn:1">1</a></sup><sup>,</sup><sup id="fnref:2"><a class="footnote-ref" href="#fn:2">2</a></sup>. In this case, each gene within a sequence is assigned to a taxon through alignment to reference genes, and the confidence of these assignments (quantified as alignment scores) is used as weights to determine a representative taxon for the entire sequence.</p>
<h2 id="retrieving-taxids-from-taxa-names">Retrieving TaxIds from taxa names<a class="headerlink" href="#retrieving-taxids-from-taxa-names" title="Permanent link">#</a></h2>
<p>In many cases, you might have the names of taxa of interest but not their corresponding TaxIds. <code>taxopy</code> provides the <a class="autorefs autorefs-internal" title="            taxopy.taxid_from_name" href="../reference/#taxopy.taxid_from_name"><code>taxid_from_name</code></a> function, which allows you to programmatically retrieve TaxIds based on taxa names. You can input a single scientific name or a list of names, and the function will return a list of TaxIds or a list of lists of TaxIds, respectively.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">taxid_from_name</span><span class="p">(</span><span class="s2">&quot;Homininae&quot;</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">))</span>
</span><span id="__span-15-2"><a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a><span class="go">[207598]</span>
</span><span id="__span-15-3"><a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">taxid_from_name</span><span class="p">([</span><span class="s2">&quot;Homininae&quot;</span><span class="p">,</span> <span class="s2">&quot;Homo sapiens&quot;</span><span class="p">],</span> <span class="n">taxdb</span><span class="p">))</span>
</span><span id="__span-15-4"><a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a><span class="go">[[207598], [9606]]</span>
</span></code></pre></div>
<p>Some taxa may share the same name, which are known as homonyms. In such cases, the returned list will include multiple TaxIds.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">taxid_from_name</span><span class="p">(</span><span class="s2">&quot;Aotus&quot;</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">))</span>
</span><span id="__span-16-2"><a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a><span class="go">[9504, 114498]</span>
</span><span id="__span-16-3"><a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">ti</span> <span class="ow">in</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">taxid_from_name</span><span class="p">(</span>
</span><span id="__span-16-4"><a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a><span class="gp">... </span>    <span class="p">[</span><span class="s2">&quot;Homininae&quot;</span><span class="p">,</span> <span class="s2">&quot;Homo sapiens&quot;</span><span class="p">,</span> <span class="s2">&quot;Aotus&quot;</span><span class="p">],</span> <span class="n">taxdb</span>
</span><span id="__span-16-5"><a id="__codelineno-16-5" name="__codelineno-16-5" href="#__codelineno-16-5"></a><span class="gp">... </span><span class="p">):</span>
</span><span id="__span-16-6"><a id="__codelineno-16-6" name="__codelineno-16-6" href="#__codelineno-16-6"></a><span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ti</span><span class="si">}</span><span class="s2">:&quot;</span><span class="p">)</span>
</span><span id="__span-16-7"><a id="__codelineno-16-7" name="__codelineno-16-7" href="#__codelineno-16-7"></a><span class="gp">... </span>    <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">ti</span><span class="p">:</span>
</span><span id="__span-16-8"><a id="__codelineno-16-8" name="__codelineno-16-8" href="#__codelineno-16-8"></a><span class="gp">... </span>        <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">taxdb</span><span class="p">)</span>
</span><span id="__span-16-9"><a id="__codelineno-16-9" name="__codelineno-16-9" href="#__codelineno-16-9"></a><span class="gp">... </span>        <span class="n">parent_taxon</span> <span class="o">=</span> <span class="n">taxon</span><span class="o">.</span><span class="n">parent</span><span class="p">(</span><span class="n">taxdb</span><span class="p">)</span>
</span><span id="__span-16-10"><a id="__codelineno-16-10" name="__codelineno-16-10" href="#__codelineno-16-10"></a><span class="gp">... </span>        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">taxon</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> (TaxId: </span><span class="si">{</span><span class="n">t</span><span class="si">}</span><span class="s2">, parent: </span><span class="si">{</span><span class="n">parent_taxon</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2">)&quot;</span><span class="p">)</span>
</span><span id="__span-16-11"><a id="__codelineno-16-11" name="__codelineno-16-11" href="#__codelineno-16-11"></a><span class="go">[207598]:</span>
</span><span id="__span-16-12"><a id="__codelineno-16-12" name="__codelineno-16-12" href="#__codelineno-16-12"></a><span class="go">Homininae (TaxId: 207598, parent: Hominidae)</span>
</span><span id="__span-16-13"><a id="__codelineno-16-13" name="__codelineno-16-13" href="#__codelineno-16-13"></a><span class="go">[9606]:</span>
</span><span id="__span-16-14"><a id="__codelineno-16-14" name="__codelineno-16-14" href="#__codelineno-16-14"></a><span class="go">Homo sapiens (TaxId: 9606, parent: Homo)</span>
</span><span id="__span-16-15"><a id="__codelineno-16-15" name="__codelineno-16-15" href="#__codelineno-16-15"></a><span class="go">[9504, 114498]:</span>
</span><span id="__span-16-16"><a id="__codelineno-16-16" name="__codelineno-16-16" href="#__codelineno-16-16"></a><span class="go">Aotus (TaxId: 9504, parent: Aotidae)</span>
</span><span id="__span-16-17"><a id="__codelineno-16-17" name="__codelineno-16-17" href="#__codelineno-16-17"></a><span class="go">Aotus (TaxId: 114498, parent: Mirbelieae)</span>
</span></code></pre></div>
<h3 id="retrieval-of-taxa-with-nearly-matching-names-though-fuzzy-search">Retrieval of taxa with nearly matching names though fuzzy search<a class="headerlink" href="#retrieval-of-taxa-with-nearly-matching-names-though-fuzzy-search" title="Permanent link">#</a></h3>
<p>When querying a <a class="autorefs autorefs-internal" title="            taxopy.TaxDb" href="../reference/#taxopy.TaxDb"><code>TaxDb</code></a> using a taxon name, you can enable fuzzy search by setting the <code>fuzzy</code> parameter of <a class="autorefs autorefs-internal" title="            taxopy.taxid_from_name" href="../reference/#taxopy.taxid_from_name"><code>taxid_from_name</code></a> to <code>True</code>. This allows the function to find taxa with names similar, but not identical, to the query string(s).</p>
<p>For a practical use case of this feature, consider the <a href="https://gtdb.ecogenomic.org/">GTDB</a> taxonomy. In GTDB <a href="https://gtdb.ecogenomic.org/faq#why-do-some-family-and-higher-rank-names-end-with-an-alphabetic-suffix">some taxa have suffixes appended to their names</a> because they are either not monophyletic in the GTDB reference tree or have unstable placements between different releases. By using fuzzy searches, you can find all the TaxIds representing a given taxon, such as <em>Myxococcota</em>, without needing to know in advance if any suffixes are appended to the name.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://github.com/shenwei356/gtdb-taxdump/releases/download/v0.5.0/gtdb-taxdump-R220.tar.gz&quot;</span>
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">gtdb_taxdb</span> <span class="o">=</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">TaxDb</span><span class="p">(</span><span class="n">taxdump_url</span><span class="o">=</span><span class="n">url</span><span class="p">)</span><span class="c1"># (1)!</span>
</span><span id="__span-17-3"><a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">taxid_from_name</span><span class="p">(</span><span class="s2">&quot;Myxococcota&quot;</span><span class="p">,</span> <span class="n">gtdb_taxdb</span><span class="p">,</span> <span class="n">fuzzy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
</span><span id="__span-17-4"><a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a><span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">gtdb_taxdb</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-17-5"><a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a><span class="go">Myxococcota_A</span>
</span><span id="__span-17-6"><a id="__codelineno-17-6" name="__codelineno-17-6" href="#__codelineno-17-6"></a><span class="go">Myxococcota</span>
</span></code></pre></div>
<ol>
<li>This custom GTDB taxdump was generated by Wei Shen using the <a href="https://bioinf.shenwei.me/taxonkit/">TaxonKit</a> toolkit and is available for download from <a href="https://github.com/shenwei356/gtdb-taxdump">GitHub</a>.</li>
</ol>
<p>You can adjust the minimum similarity threshold between the query string(s) and the matches in the database using the <code>score_cutoff</code> parameter, which determines how closely a name must match a query string to be considered a valid result. The default value is <code>0.9</code>, but you can lower this threshold to find matches that are less similar to the queries.</p>
<div class="language-pycon highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">taxopy</span><span class="o">.</span><span class="n">taxid_from_name</span><span class="p">(</span>
</span><span id="__span-18-2"><a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a><span class="gp">... </span>    <span class="s2">&quot;Myxococcota&quot;</span><span class="p">,</span> <span class="n">gtdb_taxdb</span><span class="p">,</span> <span class="n">fuzzy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">score_cutoff</span><span class="o">=</span><span class="mf">0.7</span>
</span><span id="__span-18-3"><a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a><span class="gp">... </span><span class="p">):</span>
</span><span id="__span-18-4"><a id="__codelineno-18-4" name="__codelineno-18-4" href="#__codelineno-18-4"></a><span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">taxopy</span><span class="o">.</span><span class="n">Taxon</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">gtdb_taxdb</span><span class="p">)</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
</span><span id="__span-18-5"><a id="__codelineno-18-5" name="__codelineno-18-5" href="#__codelineno-18-5"></a><span class="go">Myxococcales</span>
</span><span id="__span-18-6"><a id="__codelineno-18-6" name="__codelineno-18-6" href="#__codelineno-18-6"></a><span class="go">Myxococcota_A</span>
</span><span id="__span-18-7"><a id="__codelineno-18-7" name="__codelineno-18-7" href="#__codelineno-18-7"></a><span class="go">Myxococcus</span>
</span><span id="__span-18-8"><a id="__codelineno-18-8" name="__codelineno-18-8" href="#__codelineno-18-8"></a><span class="go">Myxococcia</span>
</span><span id="__span-18-9"><a id="__codelineno-18-9" name="__codelineno-18-9" href="#__codelineno-18-9"></a><span class="go">Myxococcota</span>
</span><span id="__span-18-10"><a id="__codelineno-18-10" name="__codelineno-18-10" href="#__codelineno-18-10"></a><span class="go">Myxococcaceae</span>
</span></code></pre></div>
<div class="footnote">
<hr />
<ol>
<li id="fn:1">
<p>Von Meijenfeldt, F. A. B., Arkhipova, K., Cambuy, D. D., Coutinho, F. H. &amp; Dutilh, B. E. "<a href="https://doi.org/10.1186/s13059-019-1817-x">Robust taxonomic classification of uncharted microbial sequences and bins with CAT and BAT</a>". <em>Genome Biology</em> <strong>20</strong>, 217 (2019).&#160;<a class="footnote-backref" href="#fnref:1" title="Jump back to footnote 1 in the text">&#8617;</a></p>
</li>
<li id="fn:2">
<p>Mirdita, M., Steinegger, M., Breitwieser, F., Sding, J. &amp; Levy Karin, E. "<a href="https://doi.org/10.1093/bioinformatics/btab184">Fast and sensitive taxonomic assignment to metagenomic contigs</a>". <em>Bioinformatics</em> <strong>37</strong>, 30293031 (2021).&#160;<a class="footnote-backref" href="#fnref:2" title="Jump back to footnote 2 in the text">&#8617;</a></p>
</li>
</ol>
</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.code.annotate", "content.code.copy", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.tabs.sticky", "toc.integrate"], "search": "../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
        <script src="../js/mermaid.mjs" type="module"></script>
      
    
  </body>
</html>